src/eradiate_plugins/integrators/piecewise_volpath.cpp

```diff
@@ -0,0 +1,443 @@
+        Interaction3f last_scatter_event = dr::zeros<Interaction3f>();
+        Float last_scatter_direction_pdf = 1.f;
+
+        /* Set up a Dr.Jit loop (optimizes away to a normal loop in scalar mode,
+           generates wavefront or megakernel renderer based on configuration).
+           Register everything that changes as part of the loop here */
+        dr::Loop<Mask> loop("Volpath integrator",
+                            /* loop state: */ active, depth, ray, throughput,
+                            result, si, mei, medium, eta, last_scatter_event,
+                            last_scatter_direction_pdf, needs_intersection,
+                            specular_chain, valid_ray, sampler);
+
+        while (loop(active)) {
+            // ----------------- Handle termination of paths ------------------
+            // Russian roulette: try to keep path weights equal to one, while accounting for the
+            // solid angle compression at refractive index boundaries. Stop with at least some
+            // probability to avoid  getting stuck (e.g. due to total internal reflection)
+            active &= dr::any(dr::neq(unpolarized_spectrum(throughput), 0.f));
+            Float q = dr::minimum(dr::max(unpolarized_spectrum(throughput)) * dr::sqr(eta), .95f);
+            Mask perform_rr = (depth > (uint32_t) m_rr_depth);
+            active &= sampler->next_1d(active) < q || !perform_rr;
+            dr::masked(throughput, perform_rr) *= dr::rcp(dr::detach(q));
+
+            active &= depth < (uint32_t) m_max_depth;
+            if (dr::none_or<false>(active))
+                break;
+
+            // ----------------------- Sampling the RTE -----------------------
+            Mask active_medium  = active && dr::neq(medium, nullptr);
+            Mask active_surface = active && !active_medium;
+            Mask act_null_scatter = false, act_medium_scatter = false,
+                 escaped_medium = false;
+
+            // If the medium does not have a spectrally varying extinction,
+            // we can perform a few optimizations to speed up rendering
+            Mask is_spectral = active_medium;
+            Mask not_spectral = false;
+            if (dr::any_or<true>(active_medium)) {
+                is_spectral &= medium->has_spectral_extinction();
+                not_spectral = !is_spectral && active_medium;
+            }
+
+            if (dr::any_or<true>(active_medium)) {
+                Float tr = dr::zeros<Float>();
+                Float pdf = dr::zeros<Float>();
+                
+                Mask intersect = needs_intersection && active_medium;
+                if (dr::any_or<true>(intersect))
+                    dr::masked(si, intersect) = scene->ray_intersect(ray, intersect);
+                needs_intersection &= !active_medium;
+                
+                std::tie( mei, tr, pdf ) = medium->sample_interaction_real(ray, si, sampler->next_1d(active_medium), channel, active_medium);
```

Comment:
Please make a clang-format pass on the file to apply the recommended formatting.